More env support

kubiko opened the pull request:
Add support to define tee supplicant paths at the run time using env variables.
`OP_TEE_FS_PARENT_PATH`, `OP_TEE_LIB`, and `OP_TEE_PLUGIN_LOAD_PATH`

When trying to run tee supplicant process in confinement, some critical directories needs to be defined dynamically during runtime, rather than at the build time.
This PR enables this run time scenario.

jforissier commented:
Hello @kubiko,

Thanks for the patches. Implementing dynamic paths sounds reasonable to me, but I have a feeling it would be better to use command line arguments rather than environment variables.

One reason is that it might be inconvenient to set the variables when starting the TEE supplicant. For example in an init script such as this one: https://github.com/OP-TEE/build/blob/3.16.0/br-ext/package/optee_client_ext/S30optee#L43-L44, the BusyBox implementation of `start-stop-daemon` doesn't support the `-e` argument to set a variable.

kubiko commented:
Hi @jforissier 

That is fair ask, let me rework the patches to address this. I need to get familiar how to pass values within supplicant.
I picked env approach to complement to default behaviour as it's least intrusive to code.
cheers
Ondra

kubiko commented:
@jforissier checking code, passing `OP_TEE_LIB` and `OP_TEE_PLUGIN_LOAD_PATH` as command line arguments.
It might be more tricky for `OP_TEE_FS_PARENT_PATH` as it's running in separate thread. Only way to get it to the [process_one_request](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supplicant.c#L596) function is to modify [thread_arg](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supplicant.c#L84), and then passing it to the [tee_supp_fs_process](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supp_fs.h#L34) as extra param.

Before implementing it, is that what you had in mind?
And if so. Do you want env and command line arguments options to coexist or to replace env option with command line arguments?

jforissier commented:
> @jforissier checking code, passing `OP_TEE_LIB` and `OP_TEE_PLUGIN_LOAD_PATH` as command line arguments. It might be more tricky for `OP_TEE_FS_PARENT_PATH` as it's running in separate thread. Only way to get it to the [process_one_request](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supplicant.c#L596) function is to modify [thread_arg](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supplicant.c#L84), and then passing it to the [tee_supp_fs_process](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/tee_supp_fs.h#L34) as extra param.

Much simpler than that: you can use global variables. `main()` would parse the command line and set the variables accordingly.

> Before implementing it, is that what you had in mind? And if so. Do you want env and command line arguments options to coexist or to replace env option with command line arguments?

I suggest not using environment variables, only the command line.

kubiko commented:
Hi @jforissier
sorry for the delay, I was held by other tasks.
I have forced pushed refactored version as per your suggestion.
env variables are now removed and it is support by arguments.
I have refactored arg parsing, as original way was quite basic and would not really scale.
cheers
Ondra

jforissier commented on tee-supplicant/src/tee_supplicant.h:
```diff
@@ -38,6 +38,15 @@
+    const char *fs_parent_path;
+};
+
+struct tee_supplicant_params supplicant_params;
```
```suggestion
struct tee_supplicant_params supplicant_params = {
	.ta_dir = "optee_armtz",
	.plugin_load_path = TEE_PLUGIN_LOAD_PATH,
	.fs_parent_path  = TEE_FS_PARENT_PATH,
};
```

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -691,16 +692,51 @@ int main(int argc, char *argv[])
+        {"plugin-path",     required_argument, 0, 'p'},
+        {0, 0, 0, 0 }
+    };
+	// set default values
+	supplicant_params.ta_dir = "optee_armtz";
+	supplicant_params.plugin_load_path = TEE_PLUGIN_LOAD_PATH;
+	supplicant_params.fs_parent_path = TEE_FS_PARENT_PATH;
```
No need to set defaults here.

kubiko replied:
Removed.

jforissier replied:
The lines are still there :open_mouth:

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -691,16 +692,51 @@ int main(int argc, char *argv[])
-		else
-			dev = argv[i];
+    static struct option long_options[] = {
+        // long name      | has argument  | flag | short value
```
Prefer C-style comments `/* long name      | has argument  | flag | short value */

kubiko replied:
fixed

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -691,16 +692,51 @@ int main(int argc, char *argv[])
+	supplicant_params.fs_parent_path = TEE_FS_PARENT_PATH;
+
+	int long_index =0;
+    while ((opt = getopt_long(argc, argv, "hdf:t:p:", long_options, &long_index )) != -1) {
```
Does Android have `getopt_long()`?

kubiko replied:
It does seems so. This is in JellyBean, that ages old...
[bionic/libc/unistd/getopt_long.c](https://android.googlesource.com/platform/bionic/+/refs/heads/jb-release/libc/unistd/getopt_long.c#518)

jforissier replied:
Alright, that's good 👍

jforissier reviewed (commented):
Thanks for the update. Overall I agree with the approach. A few issues however... see below.

jforissier commented:
Thanks for the update, LGTM. Please squash the fixup commits into the original ones. If I were you, I would make this two commits only because splitting each config value goes a bit too far IMO:
```
tee-supplicant: add definition of global connfig parameters
tee-supplicant: fs: use fs parent path from global config
tee-supplicant: use optee_armtz path from global tee parameters
tee-supplicant: plugin: use global config for plugin load path

=> tee-supplicant: introduce struct tee_supplicant_params for global config

tee-supplicant: refactor argument parsing in main()
tee-supplicant: update help info

=> tee-supplicant: refactor argument parsing in main()
```
But it is up to you.

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -98,10 +99,14 @@ struct param_value {
 static void *thread_main(void *a);
 
+struct tee_supplicant_params supplicant_params = {
+    .ta_dir = "optee_armtz",
```
indent with a tab

kubiko replied:
fixed

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -694,16 +703,47 @@ int main(int argc, char *argv[])
-			return usage(EXIT_SUCCESS);
-		else
-			dev = argv[i];
+    static struct option long_options[] = {
```
indent with tabs

kubiko replied:
fixed

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -694,16 +703,47 @@ int main(int argc, char *argv[])
+				return usage(EXIT_FAILURE);
+		}
+	}
+	// check for non option argument, which is device name
```
for consistency, use `/* ... */` for single line comments.
ditto at line 742.

kubiko replied:
fixed

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -481,9 +485,14 @@ static int get_dev_fd(uint32_t *gen_caps)
+	fprintf(stderr, "\t-d, --daemonize: run as a daemon (fork after successful "
 			"initialization)\n");
+	fprintf(stderr, "\t-f, --fs-parent-path: secure fs parent path\n");
+	fprintf(stderr, "\t-t, --ta-dir: path with TAs (optee_armtz)\n");
```
Actually should rather be something like:
```c
	fprintf(stderr, "\t-t, --ta-dir: TAs dirname under %s\n", TEEC_LOAD_PATH);
```

kubiko replied:
updated as per suggestion

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -53,6 +53,7 @@
 #include <tee_supp_fs.h>
 #include <tee_supplicant.h>
 #include <unistd.h>
+#include <getopt.h>
```
move up to line 34, following the alphabetical ordering.

kubiko replied:
fixed

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -697,16 +703,51 @@ int main(int argc, char *argv[])
+		{"fs-parent-path",  required_argument, 0, 'f'},
+		{"ta-dir",          required_argument, 0, 't'},
+		{"plugin-path",     required_argument, 0, 'p'},
+		{0, 0, 0, 0 }
```
add a space char after opening brace ` {` and before closing brace `}` for each of the lines above.

kubiko replied:
fixed

kubiko commented:
> Thanks for the update, LGTM. Please squash the fixup commits into the original ones. If I were you, I would make this two commits only because splitting each config value goes a bit too far IMO:
> 
> ```
> tee-supplicant: add definition of global connfig parameters
> tee-supplicant: fs: use fs parent path from global config
> tee-supplicant: use optee_armtz path from global tee parameters
> tee-supplicant: plugin: use global config for plugin load path
> 
> => tee-supplicant: introduce struct tee_supplicant_params for global config
> 
> tee-supplicant: refactor argument parsing in main()
> tee-supplicant: update help info
> 
> => tee-supplicant: refactor argument parsing in main()
> ```
> 
> But it is up to you.

Hi @jforissier
I'm all for clean history, commits are squashed now.
I did not squash them earlier, to make it easier for reviewers to track PR progress

etienne-lms commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -697,16 +703,51 @@ int main(int argc, char *argv[])
+		fprintf(stderr, "Using device %s.\n", argv[optind]);
+		dev = argv[optind];
+		/* check that we do not have too many arguments */
+		if (argv[optind+1]) {
```
```patch
-		if (argv[optind+1]) {
+		if (argv[optind + 1]) {

kubiko replied:
fixed

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -481,9 +485,14 @@ static int get_dev_fd(uint32_t *gen_caps)
 {
-	fprintf(stderr, "Usage: tee-supplicant [-d] [<device-name>]\n");
-	fprintf(stderr, "       -d: run as a daemon (fork after successful "
+	fprintf(stderr, "Usage: tee-supplicant [options] [<device-name>]\n");
+	fprintf(stderr, "\t-h, --help: this help\n");
+	fprintf(stderr, "\t-d, --daemonize: run as a daemon (fork after successful "
 			"initialization)\n");
+	fprintf(stderr, "\t-f, --fs-parent-path: secure fs parent path\n");
+	fprintf(stderr, "\t-t, --ta-dir: TAs dirname under %s\n", TEEC_LOAD_PATH);
+	fprintf(stderr, "\t-p, --plugin-path: plugin load path\n");
```
I suggest adding the default values in brackets:
```suggestion
	fprintf(stderr, "Usage: tee-supplicant [options] [<device-name>]\n");
	fprintf(stderr, "\t-h, --help: this help\n");
	fprintf(stderr, "\t-d, --daemonize: run as a daemon (fork after successful "
			"initialization)\n");
	fprintf(stderr, "\t-f, --fs-parent-path: secure fs parent path [%s]\n",
			supplicant_params.fs_parent_path);
	fprintf(stderr, "\t-t, --ta-dir: TAs dirname under %s [%s]\n", TEEC_LOAD_PATH,
			supplicant_params.ta_dir);
	fprintf(stderr, "\t-p, --plugin-path: plugin load path [%s]\n",
			supplicant_params.plugin_load_path);
```

kubiko replied:
updated as per suggestion

kubiko commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -688,7 +694,7 @@ int main(int argc, char *argv[])
 	char *dev = NULL;
 	int e = 0;
-	int i = 0;
+	int opt= 0;
```
fixed

kubiko commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -697,16 +703,51 @@ int main(int argc, char *argv[])
+	supplicant_params.plugin_load_path = TEE_PLUGIN_LOAD_PATH;
+	supplicant_params.fs_parent_path = TEE_FS_PARENT_PATH;
+
+	int long_index =0;
```
moved to the function entry

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -688,7 +696,8 @@ int main(int argc, char *argv[])
 	char *dev = NULL;
 	int e = 0;
-	int i = 0;
+	int long_index =0;
```
`= 0;` (missing space)

kubiko replied:
fixed

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -697,16 +706,50 @@ int main(int argc, char *argv[])
+	supplicant_params.plugin_load_path = TEE_PLUGIN_LOAD_PATH;
+	supplicant_params.fs_parent_path = TEE_FS_PARENT_PATH;
+
+	while ((opt = getopt_long(argc, argv, "hdf:t:p:", long_options, &long_index )) != -1) {
```
wrap lines < 80 characters if possible

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -697,16 +706,50 @@ int main(int argc, char *argv[])
+		dev = argv[optind];
+		/* check that we do not have too many arguments */
+		if (argv[optind + 1]) {
+			fprintf(stderr, "Too many arguments passed: extra argument: %s.\n", argv[optind+1]);
```
wrap line

kubiko replied:
fixed

jforissier reviewed (approved):
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

kubiko commented:
Hi @jforissier @etienne-lms 
I have update patch, there were two instances when logs were still using static path
```
-               IMSG("could not open directory %s", TEE_PLUGIN_LOAD_PATH);
+               IMSG("could not open directory %s", supplicant_params.plugin_load_path);
```
```
-                               TEE_FS_PARENT_PATH);
+                               tee_fs_root);
```

jforissier reviewed (commented):
Commenting on commit "tee-supplicant: fix pugin loading logic":
- In the commit subject: s/pugin/plugin/
- In the commit description: `readdir()` and `load_plugin()` (add parentheses to denote a function)
- In the commit description: s/itseld/itself/

With that fixed:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

kubiko commented:
> viewed-by: Jerome Forissier <jerome@forissier.org>

do you want this as separate PR, or keep it part of this PR?

jforissier commented:
> @jforissier @etienne-lms white testing the changes, I have also found error in the plugin loading logic. `d_name` variable in `struct dirent` returned by `readdir(dir)` contains filename relative to the searched directory, not a full path.
> 
> see this condition in [plugin.c](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/plugin.c#L127) it compares against `.` and `..` which are clearly not full paths. 

`.` and `..` are excluded because they cannot be not plugin files.

> [load_plugin](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/plugin.c#L138) takes this filename and uses it directly to load the library with [dlopen](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/plugin.c#L79) Not quite sure if or how this ever worked, but passed filename should be extended with the dir path.

Maybe it worked if `tee-supplicant` current working directory happened to be the same as `TEE_PLUGIN_LOAD_PATH`.

Ping @anisyanka.

jforissier commented:
> do you want this as separate PR, or keep it part of this PR?

Let's keep everything in this PR since the discussion has begun already.

anisyanka commented:
>> [load_plugin](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/plugin.c#L138) takes this filename and uses it directly to load the library with [dlopen](https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/src/plugin.c#L79) Not quite sure if or how this ever worked, but passed filename should be extended with the dir path.

>Maybe it worked if tee-supplicant current working directory happened to be the same as TEE_PLUGIN_LOAD_PATH.
Ping @anisyanka.


Hi @kubiko and @jforissier , sorry for the delay. I have just seen this topic.

`dlopen()` looks for shared objects either in $LD_LIBRARY_PATH env or in search path hardcoded in an executable file with the help of `-rpath` linker option.
For the supplicant plugin framework the second approach was chosen: https://github.com/OP-TEE/optee_client/blob/master/tee-supplicant/Makefile#L84
It seems there is no mistake in plugins load logic.

anisyanka commented:
Sounds good, agree.
Let's move "tee-supplicant: fix plugin loading logic" to another PR.

kubiko commented:
@jforissier commits are now updated with `Reviewed-by:`
do you want me to open another PR with plugin load logic fix?

jforissier commented:
> do you want me to open another PR with plugin load logic fix?

Feel free to do so.

kubiko commented:
> @anisyanka thanks for your reply.
> 
> > `dlopen()` looks for shared objects either in $LD_LIBRARY_PATH env or in search path hardcoded in an executable file with the help of `-rpath` linker option.
> 
> True, but `dlopen()` will also accept absolute paths in which case `$LD_LIBRARY_PATH` and `-rpath` are irrelevant (well, at least to load the file given to `dlopen()` -- dependencies are another story). Giving an absolute path to the plugin seems easier to understand and safer too because there would be no redirection possible via `$LD_LIBRARY_PATH`.
> 
> That being said I think such a change belongs in another PR. I recommend simply dropping the patch "tee-supplicant: fix plugin loading logic" for now. What do you think?

I agree with @jforissier here. We scan defined folder for the list of plugins to load, but actual plugin can then come from totally different place, if one plays with `LD_LIBRARY_PATH`.
Plugin dependencies is fair concern, but this should not have affect on the actual plugin file to load.
As there is agreement to move this to separate PR, I'm opening [new PR](https://github.com/OP-TEE/optee_client/pull/300) for this.
